rust/hyperlane-core/src/types/channel.rs

```diff
@@ -0,0 +1,50 @@
+use derive_new::new;
+use tokio::sync::broadcast::{Receiver, Sender};
+
+/// Multi-producer, multi-consumer channel
+pub struct MpmcChannel<T> {
+    sender: Sender<T>,
+    receiver: MpmcReceiver<T>,
+}
+
+impl<T: Clone> MpmcChannel<T> {
+    /// Creates a new `MpmcChannel` with the specified capacity.
+    ///
+    /// # Arguments
+    ///
+    /// * `capacity` - The maximum number of messages that can be buffered in the channel.
+    pub fn new(capacity: usize) -> Self {
+        let (sender, receiver) = tokio::sync::broadcast::channel(capacity);
+        Self {
+            sender: sender.clone(),
+            receiver: MpmcReceiver::new(sender, receiver),
+        }
+    }
+
+    /// Returns a clone of the sender end of the channel.
+    pub fn sender(&self) -> Sender<T> {
+        self.sender.clone()
+    }
+
+    /// Returns a clone of the receiver end of the channel.
+    pub fn receiver(&self) -> MpmcReceiver<T> {
+        self.receiver.clone()
+    }
+}
+
+/// Clonable receiving end of a multi-producer, multi-consumer channel
+#[derive(Debug, new)]
+pub struct MpmcReceiver<T> {
+    sender: Sender<T>,
+    /// The receiving end of the channel.
+    pub receiver: Receiver<T>,
+}
+
+impl<T> Clone for MpmcReceiver<T> {
+    fn clone(&self) -> Self {
+        Self {
+            sender: self.sender.clone(),
+            receiver: self.sender.subscribe(),
```

Comment:
Curious if you considered implications of creating a new subscriber each time we clone https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html#lagging

I think it's probably fine, in the worst case we'll have some subscribers that never read from the receiver, so at most we'll keep 1000 operations in the channel, and they'll start to be removed from the channel as new ones come in. Does make me wonder if we should have a lower capacity than 1000, but a few hundred Kb stuck around at most isn't a big deal

Reply:
hadn't considered this exact edge case. I think that if there are any redundant receivers they're only temporary and are dropped eventually. Also don't think it's a big deal, would rather have very readable code for now